kerchunk/df.py

```diff
@@ -123,242 +128,102 @@ def refs_to_dataframe(
         Passed to fsspec when for writing the parquet.
     row_group_size : int
         Number of references to store in each reference file (default 1000)
-    compression : str
-        Compression information to pass to parquet engine, default is zstd.
-    engine : {'fastparquet', 'pyarrow'}
-        Library to use for writing parquet files.
     **kwargs : dict
         Additional keyword arguments passed to parquet engine of choice.
     """
-    if not os.path.exists(url):
-        os.makedirs(url)
+    consolidated = True  # not even an argument, let's just use it
     if "refs" in refs:
         refs = refs["refs"]
-    _write_json(
-        os.path.join(url, ".row_group_size"), dict(row_group_size=row_group_size)
-    )
+    if consolidated:
+        # because all metadata are embedded
+        refs = zarr.convenience.consolidate_metadata(refs)
+
+    # write into .zmetadata at top level, one fewer read on access
+    refs[".row_group_size"] = '{"row_group_size": %i}' % row_group_size
+    # _write_json(
+    #    "/".join([url, ]), dict(row_group_size=row_group_size),
+    #    storage_options=storage_options
+    # )
+
+    fs, _ = fsspec.core.url_to_fs(url)
+    fs.makedirs(url, exist_ok=True)
     fields = get_variables(refs, consolidated=consolidated)
     for field in fields:
-        field_path = os.path.join(url, field)
+        field_path = "/".join([url, field])
         if field.startswith("."):
             # zarr metadata keys (.zgroup, .zmetadata, etc)
-            _write_json(field_path, refs[field])
-        else:
-            if not os.path.exists(field_path):
-                os.makedirs(field_path)
-            # Read the variable zarray metadata to determine number of chunks
-            zarray = ujson.loads(refs[f"{field}/.zarray"])
-            chunk_sizes = np.ceil(
-                np.array(zarray["shape"]) / np.array(zarray["chunks"])
+            _write_json(field_path, refs[field], storage_options=storage_options)
+            continue
+
+        fs.makedirs(field_path, exist_ok=True)
+        # Read the variable zarray metadata to determine number of chunks
+        zarray = ujson.loads(refs[f"{field}/.zarray"])
+        chunk_sizes = np.ceil(np.array(zarray["shape"]) / np.array(zarray["chunks"]))
+        if chunk_sizes.size == 0:
+            chunk_sizes = np.array([0])
+        nchunks = int(np.product(chunk_sizes))
+        extra_rows = row_group_size - nchunks % row_group_size
+        output_size = nchunks + extra_rows
+
+        paths = np.full(output_size, np.nan, dtype="O")
+        offsets = np.zeros(output_size, dtype="int64")
+        sizes = np.zeros(output_size, dtype="int64")
+        raws = np.full(output_size, np.nan, dtype="O")
+        nmissing = 0
+
+        for metakey in [".zarray", ".zattrs"]:
+            # skip when consolidated?
```

Comment:
Actually thinking about this some more, I think it might actually make sense to just load these directly from `.zmetadata` since having to parse it isn't really a big deal. I'll go ahead and give this a try.